### PR TITLE
Add support for exp and log (with approx) on additive shared tensors

### LINKED ISSUE #2537: Exponential implementation in SMPC
**Suggestion to implement exp in SMPC**

The interest of this is to allow computation of functions which derive of exponentials like sigmoid.

Let's say you share the tensor x in x0 and x1. You want to compute exp(x).
But exp(x) = exp(x0 + x1) = exp(x0) * exp(x1).
So you can compute locally exp(x0) and exp(x1), and **additively share** these values, perform an encrypted mul and then you obtain shares of exp(x0) * exp(x1) = exp(x).

### PR DESCRIPTION
**Pull request by LaRiffle**
This PR extends the set of operations supported in Additive Sharing using approximations

- [x] Exponential 
- [x] Logarithm
- [x] Inverse
- [x] Sigmoid

Fix #2537

### CODE REVIEW

File: syft/frameworks/torch/tensors/interpreters/precision.py [hunk 1]
```diff
@@ -182,8 +183,8 @@ def __iadd__(self, other):
     def sub(self, _self, other):
         """Subtracts a fixed precision tensor from another one.
         """
-        if isinstance(other, int):
-            scaled_int = other * self.base ** self.precision_fractional
+        if isinstance(other, (int, float)):
+            scaled_int = int(other * self.base ** self.precision_fractional)
```
Comment (robert-wagner) [inline syft/frameworks/torch/tensors/interpreters/precision.py]: Should this be an int or a long?

File: syft/frameworks/torch/tensors/interpreters/precision.py [hunk 2]
```diff
@@ -443,6 +447,70 @@ def matmul(self, *args, **kwargs):
     __matmul__ = matmul
     mm = matmul
 
+    # Approximations:
+    def exp(self, iterations=8):
```
[the remark this answers is not in the document]
Comment (robert-wagner) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: Same with both below functions

[the remark this answers is not in the document]
Comment (jvmancuso) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: These should also have citations, since the inspiration comes from CrypTen. I'd say it's not substantial enough to require that the license be copied, so a simple online comment + link should do

File: syft/frameworks/torch/tensors/interpreters/precision.py [hunk 3]
```diff
@@ -202,6 +203,9 @@ def sub(self, _self, other):
 
     __sub__ = sub
 
+    def __rsub__(self, other):
```
[the remark this answers is not in the document]
Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: Not exactly because `__rsub__(self, other)` is called after `other - self` failed, typically because other was an integer. So we need to reverse arguments.

File: syft/frameworks/torch/tensors/interpreters/precision.py (hunk 1 above)
Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: float can be supported too

File: syft/frameworks/torch/tensors/interpreters/precision.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: Indeed, but I think this is a good temporary place until polynomial is ready to receive them and we have a clear strategy to integrate Poly & FixedP.
Also, I've added references to the repo where I analyse, test and adapt those approximations suggested by crypten.

File: syft/frameworks/torch/tensors/interpreters/precision.py [hunk 4]
```diff
@@ -443,6 +451,73 @@ def matmul(self, *args, **kwargs):
...
+
+        Args:
+            iterations (int): number of iterations for limit approximation
+
+        Ref: https://github.com/LaRiffle/approximate-models
+        """
+        return (1 + self / 2 ** iterations) ** (2 ** iterations)
+
+    def sigmoid(self):
+        """
```
Comment (andrelmfarias) [inline syft/frameworks/torch/tensors/interpreters/precision.py]: Shouldn't we replace the sigmoid approximation by the mathematical definition using the approximated `exp`? Or do you think we lose precision by doing that?

I think we should at least compare the differences between both...

**Comment by andrelmfarias** [comment]
Don't you think we should also add some efficiency tests in `test/efficiency_tests`, as I did for `tanh` and `sigmoid`?

File: syft/frameworks/torch/tensors/interpreters/precision.py (hunk 4 above)
Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: Definitely! I'll make a comparison as part of my lib to see what's best, thansk for the suggestion

**Comment by LaRiffle** [comment]
Yes, good idea!

File: syft/frameworks/torch/tensors/interpreters/precision.py (hunk 4 above)
Comment (andrelmfarias) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: You're welcome!

Maybe also do a time performance comparison, to see what are the trade-offs between both

**Comment by iamtrask** [comment]
Super excited about this!

File: syft/frameworks/torch/tensors/interpreters/precision.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (iamtrask) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: Thank you for doing that @LaRiffle and @jvmancuso

File: test/torch/tensors/test_precision.py [hunk 5]
```diff
@@ -326,6 +384,118 @@ def test_torch_dot(workers):
...
+            )
+            r_sh = t_sh.exp()
+            r = r_sh.get().float_prec()
+            t = t.exp()
+            diff = (r - t).abs()
+            norm = (r + t) / 2
+            cumsum += diff / (tolerance * norm)
+
+        cumsum /= 10
+        print(cumsum)
```
Comment (Jasopaum) [inline test/torch/tensors/test_precision.py]: This can be removed :)

File: test/torch/tensors/test_precision.py [hunk 6]
```diff
@@ -326,6 +384,118 @@ def test_torch_dot(workers):
...
+    Test the approximate exponential with different tolerance depending on
+    the precision_fractional considered
+    """
+    alice, bob, james = workers["alice"], workers["bob"], workers["james"]
+
+    fix_prec_tolerance = {3: 20 / 100, 4: 5 / 100, 5: 5 / 100}
+
+    for prec_frac, tolerance in fix_prec_tolerance.items():
+        cumsum = torch.zeros(5)
+        for i in range(10):
```
Comment (Jasopaum) [inline test/torch/tensors/test_precision.py]: Are you doing this because the approximation results vary significantly depending on the sharing?
If so, do you know when "bad" cases occur and how bad they are?

File: syft/frameworks/torch/tensors/interpreters/precision.py [hunk 7]
```diff
@@ -443,6 +454,104 @@ def matmul(self, *args, **kwargs):
...
+                    sigmoid(x) = 1 / (1 + exp(-x))
+                "maclaurin": Use the Maclaurin / Taylor approximation, with polynomial
+                    interpolation of degree 5 over [-8,8]
+                    NOTE: This method is faster but not as precise as "exp"
+                    Ref: https://mortendahl.github.io/2017/04/17/private-deep-learning-with-mpc/#approximating-sigmoid
+        """
+        one = self * 0 + 1
+
+        if method == "exp":
+            result = one / (1 + (self * -1).exp())
```
Comment (Jasopaum) [inline syft/frameworks/torch/tensors/interpreters/precision.py]: I think we might have talked about it at some point but should you use `.inverse()` here?

File: test/torch/tensors/test_precision.py [hunk 8]
```diff
@@ -93,9 +81,33 @@ def test_methods_for_linear_module(method, parameter):
...
+    # Method syntax
+    x = torch.tensor([0.1, 0.2, 0.3]).fix_prec()
+
+    y = x + x
+
+    assert (y.child.child == torch.LongTensor([200, 400, 600])).all()
+    y = y.float_prec()
+
+    assert (y == torch.tensor([0.2, 0.4, 0.6])).all()
+
```
[the remark this answers is not in the document]
Comment (LaRiffle) [reply test/torch/tensors/test_precision.py]: I will remove it

File: test/torch/tensors/test_precision.py (hunk 6 above)
Comment (LaRiffle) [reply test/torch/tensors/test_precision.py]: Yes
No not really... :/

File: syft/frameworks/torch/tensors/interpreters/precision.py (hunk 7 above)
Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: Correct!
